### PR TITLE
CA-464 chunkify testing permissions for billing accounts

### PR DESCRIPTION
**Pull request by helgridly**
The quota is 300 requests per minute (see [here](https://console.cloud.google.com/apis/api/cloudbilling.googleapis.com/quotas) for some project).

If someone had 300 billing projects and the turnaround time for one IAM test is 2 seconds, we could still hit the quota. (10 requests every 2 seconds = 300/min)

If we don't like this approach we could steal [the throttler from GPAlloc](https://github.com/broadinstitute/gpalloc/blob/develop/src/main/scala/org/broadinstitute/dsde/workbench/gpalloc/util/Throttler.scala).

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

**Comment by helgridly** [comment]
@rtitle i'm not super confident on whether my use of IO here is right so i'd appreciate the sanity check :)

**Review by marctalbott — APPROVED** [review]

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala [hunk 1]
```diff
@@ -659,14 +659,28 @@ class HttpGoogleServicesDAO(
...
+      //some users have TONS of billing accounts, enough to hit quota limits.
+      //break the list of billing accounts up into chunks, and process the chunks serially.
+      //this should slow things down somewhat, though we may need to upgrade this to a throttle.
+      val accountChunks: List[Seq[BillingAccount]] = accountList.grouped(10).toList
+
+      //Iterate over each chunk.
+      val allProcessedChunks: IO[List[Seq[RawlsBillingAccount]]] = accountChunks traverse { chunk =>
+
+        //Future.sequence each chunk (i.e. run all tests in the chunk in parallel)
+        IO.fromFuture(IO(Future.sequence(chunk map { acct =>
```
Comment (rtitle) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala]: I think
```
Future.sequence(chunk map { acct => ... })
```
is equivalent to
```
chunk traverse { acct => ... }
```

**Comment by dvoet** [comment]
Have we asked google for a quota increase or a better way?

**Review by rtitle — APPROVED** [review]
The code looks good -- although making the point that even with chunking we could still hit quota without some sort of throttling/delay.

**Comment by helgridly** [comment]
@dvoet I can certainly ask for a quota increase

**Comment by dvoet** [comment]
Seems like the problem is that we have too much parallelism and that chunking limits that. Can you add something to that effect in the comments? Explain why chunking works.

**Comment by helgridly** [comment]
@dvoet what are you looking for beyond the comment that's already there? https://github.com/broadinstitute/rawls/pull/1124/files#diff-5ecf29f76a772ac794b30c2ba0a65dd7R669

**Comment by helgridly** [comment]
jenkins retest

**Comment by dvoet** [comment]
The old code did them all in parallel. New code does 10 in parallel. The comment is unclear that all elements in a chunk execute in parallel. So really the size of the chunks governs the parallelism. Chunking makes it feel batchy to me which is not the case. I just want to make it clear that the intent is to limit parallelism.

**Pull request merged**